lib/spack/spack/binary_distribution.py

```diff
@@ -0,0 +1,209 @@
+    buildinfo['buildpath'] = spack.install_path
+    buildinfo['relocate_textfiles'] = text_to_relocate
+    buildinfo['relocate_binaries']  = binary_to_relocate
+    filename = buildinfo_file_name(spec)
+    with open(filename, 'w') as outfile:
+        outfile.write(yaml.dump(buildinfo, default_flow_style=True))
+
+
+def tarball_directory_name(spec):
+    """
+    Return name of the tarball directory according to the convention
+    <os>-<architecture>/<compiler>/<package>/
+    """
+    return "%s/%s/%s" % (get_full_system_from_platform(),
+                         str(spec.compiler).replace("@", "-"),
+                         spec.name)
+
+
+def tarball_name(spec):
+    """
+    Return the name of the tarfile according to the convention
+    <os>-<architecture>-<package>-<dag_hash>.tar.gz
+    """
+    return "%s-%s-%s-%s.tar.gz" % (get_full_system_from_platform(),
+                                   spec.name,
+                                   spec.version,
+                                   spec.dag_hash())
+
+
+def tarball_path_name(spec):
+    """
+    Return the full path+name for a given spec according to the convention
+    <tarball_directory_name>/<tarball_name>
+    """
+    return os.path.join(tarball_directory_name(spec),
+                        tarball_name(spec))
+
+
+def build_tarball(spec, outdir, force=False):
+    """
+    Build a tarball from given spec and put it into the directory structure
+    used at the mirror (following <tarball_directory_name>).
+    """
+    tarfile_dir = os.path.join(outdir, tarball_directory_name(spec))
+    tarfile = os.path.join(outdir, tarball_path_name(spec))
+    if os.path.exists(tarfile):
+        if force:
+            os.remove(tarfile)
+        else:
+            tty.warn("file exists, use -f to force overwrite: %s" % tarfile)
+            return
+    if not os.path.exists(tarfile_dir):
+        mkdirp(tarfile_dir)
+
+    tar = which('tar', required=True)
+    dirname = os.path.dirname(spec.prefix)
+    basename = os.path.basename(spec.prefix)
+
+    # handle meta-data
+    cp = which("cp", required=True)
+    spec_file = os.path.join(spec.prefix, ".spack", "spec.yaml")
+    target_spec_file = tarfile + ".yaml"
+    cp(spec_file, target_spec_file)
+
+    # create info for later relocation and create tar
+    write_buildinfo_file(spec)
+    tar("--directory=%s" % dirname, "-czf", tarfile, basename)
+    tty.msg(tarfile)
+
+
+def download_tarball(package):
+    """
+    Download binary tarball for given package into stage area
+    Return True if successful
+    """
+    mirrors=spack.config.get_config('mirrors')
+    if len(mirrors) == 0:
+        tty.die("Please add a spack mirror to allow " +
+                "download of pre-compiled packages.")
+    tarball = tarball_path_name(package.spec)
+    for key in mirrors:
+        url=mirrors[key]+"/"+tarball
+        print url
+        # stage the tarball into standard place
+        stage = Stage(url, name=package.stage.path)
```

Comment:
Ok -- we can discuss more on the call.  I think it's worthwhile to have a more detailed discussion about this.

I'm not sure you want to actually fetch the hashes directly from the mirror.  You need to have a way to guarantee that the hash itself came from a secure channel.  If you trust the server to provide both a reliable checksum and a reliable package, you're vulnerable to a compromised server.  The hashes/other verification mechanism need to come along some secure channel.  The approach I mentioned of signing packages like Debian seems promising for this as it's easy to transfer a public key once and verify packages with it from then on.  

btw -- LLNL has some S3 space where we can start to put up a mirror.  CERN may have something similar -- we can talk about that on the call as well.